packages/core/src/routes/sign-in-experience/custom-ui-assets/index.ts

```diff
@@ -0,0 +1,100 @@
+import { readFile } from 'node:fs/promises';
+
+import { uploadFileGuard, maxUploadFileSize } from '@logto/schemas';
+import { generateStandardId } from '@logto/shared';
+import { object, z } from 'zod';
+
+import { EnvSet } from '#src/env-set/index.js';
+import RequestError from '#src/errors/RequestError/index.js';
+import koaGuard from '#src/middleware/koa-guard.js';
+import SystemContext from '#src/tenants/SystemContext.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getConsoleLogFromContext } from '#src/utils/console.js';
+import { buildAzureStorage } from '#src/utils/storage/azure-storage.js';
+import { getTenantId } from '#src/utils/tenant.js';
+
+import { type ManagementApiRouter, type RouterInitArgs } from '../../types.js';
+
+export default function customUiAssetsRoutes<T extends ManagementApiRouter>(
+  ...[router]: RouterInitArgs<T>
+) {
+  // TODO: Remove
+  if (!EnvSet.values.isDevFeaturesEnabled) {
+    return;
+  }
+
+  router.post(
+    '/sign-in-exp/default/custom-ui-assets',
+    koaGuard({
+      files: object({
+        file: uploadFileGuard.array().min(1),
+      }),
+      response: z.object({
+        customUiAssetId: z.string(),
+      }),
+      status: [200, 400, 500],
+    }),
+    async (ctx, next) => {
+      const { file: bodyFiles } = ctx.guard.files;
+      const file = bodyFiles[0];
+
+      assertThat(file, 'guard.invalid_input');
+      assertThat(file.size <= maxUploadFileSize, 'guard.file_size_exceeded');
+      assertThat(file.mimetype === 'application/zip', 'guard.mime_type_not_allowed');
+
+      const { experienceZipsProviderConfig } = SystemContext.shared;
+      assertThat(
+        experienceZipsProviderConfig?.provider === 'AzureStorage',
```

Comment:
```suggestion
        experienceZipsProviderConfig?.provider,
```

Reply:
technically we should not care about the specific storage provider. you can leave `downloadFile` in other providers not implemented (throw an error)

Reply:
The download function of Azure SDK returns `Promise<BlobDownloadResponseParsed>`. I feel bad writing this return type in a generic interface.

Reply:
Offline discussed, will try to improve this in [a separated task](https://linear.app/silverhand/issue/LOG-9642/improve-storage-interface-and-support-download-and-isexisted-functions).